### DISCUSSION TITLE
How to properly handle concurrent commits with scoped_session

### DISCUSSION THREAD

**Discussion by dmsfabiano** [sqlalchemy/sqlalchemy#10316] [Usage Questions] [answered]
I have a `Session` that is created through `scoped_session`, which I am using as a global variable on my orchestrator (dagster), which spawn's job's which call `my_session = Session()` to get the current session. Each jobs run's as a python subprocess.

What is the current way to isolate the `my_session.commit()` so that we do not get `Database is locked` exception? I have tried to do a `my_session.connection().in_transaction()` and `sleep` if it returns true. However, this seems to always return true

Each process has isolated data that they need to write to the db, and do not need to share data. I am using `sqlite` and I understand it does not support concurrency, but I was hoping i could use the scoped_sessions to manage when it is safe to commit

Is there any documentation with examples on when to commit? I have read through [this](https://docs.sqlalchemy.org/en/14/orm/contextual.html#unitofwork-contextual), but that doesn't elaborate on tools of where or how to handle the commit

**Comment by zzzeek** [comment, marked as answer]
the pysqlite driver does not start a transaction in the SQLite database until write operations (i.e. DML, UPDATE, INSERT, DELETE, as well as DDL, CREATE TABLE etc) occur.   so if you are getting `Database is locked` that means you have very concurrent writes going on and you possibly have processes ending / crashing leaving the database locked.   normally, two concurrent writes should have one of the writers waiting.  there might be a timeout of some kind.    using the [autocommit setting](https://docs.sqlalchemy.org/en/20/core/connections.html#setting-transaction-isolation-levels-including-dbapi-autocommit) might help to reduce the span of a single database lock, but overall would first start out with that the transactions in each worker are as short as possible and leave the database unlocked for other processes to operate.

overall concurrent writes to sqlite is a bigger issue than sqlalchemy so I'd also check out sources like https://www2.sqlite.org/cvstrac/wiki?p=DatabaseIsLocked  and https://www.sqlite.org/lockingv3.html for background on how to deal with this problem in general.     from the SQLAlchemy side, you want to be calling `.commit()` to end your transaction ASAP and possibly use autocommit isolation level.

**Reply by dmsfabiano** [reply]
This is really helpful, thank you! In regards to "two concurrent writes should have one of the writers waiting. there might be a timeout of some kind." is there a way to increase the wait time from the sqalchemy side through the session or engine?

**Reply by zzzeek** [reply]
looks like it is five seconds:  https://docs.python.org/3/library/sqlite3.html#sqlite3.connect

you would change this as:

    create_engine("sqlite://", connect_args={"timeout": 30})

**Reply by dmsfabiano** [reply]
I would try a few things out. Thanks again, do you know if the reads, also cause a lock? I tried the DB Commits without multi-processing, and they do not last more than `ms`, so the only thing I can think of is that `session.query(...)` also causes a lock?

**Reply by zzzeek** [reply]
the way that pysqlite / sqlite3 sets up transactions, reads should not cause exclusive locks.
